app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java

```diff
@@ -388,76 +383,19 @@ public void removeWatchedStreams(final boolean removePartiallyWatched) {
-
-                            if (indexInHistory < 0) {
-                                notWatchedItems.add(playlistItem);
-                            } else if (!thumbnailVideoRemoved
-                                    && playlistManager.getPlaylistThumbnail(playlistId)
-                                    .equals(playlistItem.getStreamEntity().getThumbnailUrl())) {
-                                thumbnailVideoRemoved = true;
-                            }
-                        }
-                    } else {
-                        final Iterator<StreamStateEntity> streamStatesIter = recordManager
-                                .loadLocalStreamStateBatch(playlist).blockingGet().iterator();
-
-                        while (playlistIter.hasNext()) {
-                            final PlaylistStreamEntry playlistItem = playlistIter.next();
-                            final int indexInHistory = Collections.binarySearch(historyStreamIds,
-                                    playlistItem.getStreamId());
-
-                            final boolean hasState = streamStatesIter.next() != null;
-                            if (indexInHistory < 0 ||  hasState) {
-                                notWatchedItems.add(playlistItem);
-                            } else if (!thumbnailVideoRemoved
-                                    && playlistManager.getPlaylistThumbnail(playlistId)
-                                    .equals(playlistItem.getStreamEntity().getThumbnailUrl())) {
-                                thumbnailVideoRemoved = true;
-                            }
-                        }
-                    }
-
-                    return Flowable.just(notWatchedItems, thumbnailVideoRemoved);
-                })
+        disposables.add(playlistManager.clearWatched(playlistId, removePartiallyWatched)
                 .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(flow -> {
-                    final List<PlaylistStreamEntry> notWatchedItems =
-                            (List<PlaylistStreamEntry>) flow.blockingFirst();
-                    final boolean thumbnailVideoRemoved = (Boolean) flow.blockingLast();
-
+                .subscribe(notWatchedItems -> {
                     itemListAdapter.clearStreamItemList();
                     itemListAdapter.addItems(notWatchedItems);
-                    saveChanges();
 
+                    final String currentThumbnail = playlistManager
+                            .getPlaylistThumbnail(playlistId);
+                    final boolean requireThumbnailChange = notWatchedItems.stream()
+                            .noneMatch(s -> Objects.equals(s.getStreamEntity().getThumbnailUrl(),
+                                    currentThumbnail));
 
-                    if (thumbnailVideoRemoved) {
+                    if (requireThumbnailChange) {
                         updateThumbnailUrl();
                     }
```

Comment:
I'd also move this to a separate function inside `LocalPlaylistManager`, I see it is being used in multiple places around this file.

Reply:
Like this?
```java
final List<PlaylistStreamEntry> notWatched;
if (removePartiallyWatched) {
   notWatched = streams.stream()
      .filter(s -> Collections.binarySearch(historyIds, s.getStreamId()) < 0)
      .collect(Collectors.toList());
} else {
   notWatched = streams.stream()
      .filter(s -> Collections.binarySearch(historyIds,  s.getStreamId()) < 0 || s.getProgressTime() > 0)
      .collect(Collectors.toList());
}
```